src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -0,0 +1,152 @@
+                return this.ChainedHeaderTree;
+            }
+
+            public ChainedHeader ExtendAChain(int count, ChainedHeader chainedHeader = null)
+            {
+                ChainedHeader previousHeader = chainedHeader ?? new ChainedHeader(this.Network.GetGenesis().Header, this.Network.GenesisHash, 0);
+
+                foreach (int index in Enumerable.Range(1, count))
+                {
+                    BlockHeader header  = this.Network.Consensus.ConsensusFactory.CreateBlockHeader();
+                    header.HashPrevBlock = previousHeader.HashBlock;
+                    header.Bits = previousHeader.Header.Bits - 1000; // just increase difficulty.
+                    ChainedHeader newHeader = new ChainedHeader(header, header.GetHash(), previousHeader);
+                    previousHeader = newHeader;
+                }
+
+                return previousHeader;
+            }
+
+            public List<BlockHeader> ChainedHeaderToList(ChainedHeader chainedHeader, int count)
+            {
+                List<BlockHeader> list = new List<BlockHeader>();
+
+                ChainedHeader current = chainedHeader;
+                while (count > 0)
+                {
+                    list.Add(current.Header);
+                    current = current.Previous;
+                    count--;
+                }
+
+                list.Reverse();
+
+                return list;
+            }
+
+            public bool ConnectedHeadersIsEmpty(ConnectedHeaders connectedHeaders)
+            {
+                Assert.NotNull(connectedHeaders);
+
+                return connectedHeaders.Consumed == null 
+                       && connectedHeaders.DownloadTo == null 
+                       && connectedHeaders.DownloadFrom == null;
+            }
+        }
+
+        [Fact]
+        public void ConnectHeaders_HeadersCantConnect_ShouldFail()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+
+            Assert.Throws<ConnectHeaderException>(() => chainedHeaderTree.ConnectNewHeaders(1, new List<BlockHeader>(new [] { testContext.Network.GetGenesis().Header})));
+        }
+
+        [Fact]
+        public void ConnectHeaders_NoNewHeadersToConnect_ShouldReturnNothingToDownload()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+
+            var chainTip = testContext.ExtendAChain(10);
+            chainedHeaderTree.Initialize(chainTip);
+
+            var listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 4);
+
+            var connectedHeaders = chainedHeaderTree.ConnectNewHeaders(1, listOfExistingHeaders);
+
+            Assert.True(testContext.ConnectedHeadersIsEmpty(connectedHeaders));
+        }
+
+        [Fact]
+        public void ConnectHeaders_HeadersFromTwoPeers_ShouldCreateTwoPeerTips()
```

Comment:
Similarly, this does not implement test case number 3

Reply:
It is perfectly fine to have such a test, it is just that our previous discussion was that those test that implemented so I just want to note that they were not. But both these tests here are fine, but still the other tests should be implemented.